Changes for week of Sep 16 (#167)
xregistry/spec: schema/spec.md

  specification defines a set of common [schema format names](#schema-formats)
  that MUST be used for the given formats, but applications MAY define
  extensions for other formats on their own.

>   For many schema registry use cases this attribute might be very important for

Reviewer: ```suggestion
  For many schema registry use cases this attribute is important for
```

There's already hedging in "For many schema registry use cases", so I don't think we need the "might be" (and "very" doesn't feel helpful to me - although that's a personal preference of course).

Author: done

}
```

#### `validation`

- Type: Boolean
- Description: Indicates whether or not the server will validate the Resource's
  document(s) against the specified `format` attribute. A value of `true`
  indicates that the server MUST reject any request that would cause any
  Version of this Resource to be invalid per the rules as defined by the

>   `format` specification. Note, this includes a request to set this attribute

Reviewer: I don't understand the note - is it saying "If there's a request to set this attribute to true, the server must validate the rest of the intended state of the resource document immediately"? (So you don't get to set it to true with an invalid document.)

Author: Yes, and in particular it means all Versions of the Resource need to be checked too since it seems like a lie to me to have "validation"=true AND have some Versions of the Resource be non-compliant per the "format" attribute. Would it help if I elaborated more  in the spec?

Reviewer: Definitely worth elaborating more. I think it's also worth *considering* the use case of "I had broken resources before, but anything from now on should be valid" (in terms of any *new* versions). Or is the intention that if you had broken ones and you want validation, you should delete all the broken ones first?

Author: yes, I was assuming that people would be forced to update all existing Versions before they could turn on validation. Annoying, but I think that's better than allowing for things to be in an inconsistent state.  That's why I have a bit of a reservation about allowing "old ones" to be non-compliant... how do we explain to new users of that Resource why some are invalid or even what the date is for the switch? "Required compliance" seems like the easier path. But let's discuss on the call.

Added a bit more text to make this clear.

  to `true`.

>   A value of `false` indicates that the server MUST NOT do any validation.

Reviewer: This is interesting. It feels to me like it would be useful to make "validate but don't prevent" an option - e.g. "The server mustn't reject requests causing an invalid version, but can report warnings". I don't know where it would report warnings, admittedly.

Author: As you said... how do you return warnings?

I guess one way to do it would be to have some kind of "status" attribute that's server controlled that would tell you whether or not the doc is valid. Then not just the current user would know but anyone checking the registry later too. Let's discuss because it does make things a bit more complex for users.

  If `format` is not specified, or if the value is not known by the server

>   (but it an allowable value), then the server MUST NOT perform any validation.

Reviewer: "any" validation feels like it's potentially asking for trouble. Would it still be allowed to reject incoming JSON which specified the same property multiple times, for example? (That's technically-valid JSON, but feels reasonable to reject.) Presumably it's *always* okay to reject requests which don't have valid JSON, or are broken in other ways.

Basically I'm saying that we should limit the scope of "must not perform any validation" otherwise it sounds like "the server must not reject any request at all".

Author: The intent was to let any array of bytes in. Imagine cases where people are purposely uploading invalid docs, perhaps for sharing testing (invalid) docs or they're using a format that's only valid because of the tooling they're using (like some kind of new templating language for the format).  I see two ways for users to say "let anything in"... 1) turn off validation,  2) use a format that's the server doesn't know about and therefore can't check (which includes "no format").

Reviewer: Right, thanks for the context. Presumably it's still only for the resource body though? I can't remember enough about other aspects of xRegistry requests to give concrete aspects, but I'd expect a sort of firewall between "the body of the resource, which is somewhat-opaque like CloudEvent data" and "resource metadata, authentication etc" which I'd still hope to be validated.

I'm hoping that you're able to clarify this in a more technically correct way, of course :)

Author: `Indicates whether or not the server will validate the Resource's document(s) against the specified `format` attribute`. It's the phrase "Resource's document" that's they key to me as that means it's that secondary doc associated with each Version that's checked and not the xReg metadata.  Think I need to elaborate more?

Reviewer: That's fair enough - I suspect the nature of line-by-line PR review made me less sensitive to the context :)
